### PR TITLE
compiler: fix nondeterministic behaviour

### PR DESCRIPTION
**Pull request by delitrem**
Fixes nondeterministic behaviour related to new doc system. See #8579.

### CODE REVIEW

**Comment by garazdawi** [comment]
Hello! Thanks for the fix! Would you mind adding a testcase to `lib/compiler/test/beam_doc_SUITE.erl` that tests that it works?

**Comment by garazdawi** [comment]
Also, would you mind targeting `maint` instead of `master` so that we can get a fix out before next year :)

**Comment by delitrem** [comment]
Hello and thank you very much for the feedback. I will write a testcase.

**Comment by delitrem** [comment]
Still I have no great idea how to reproduce 'bad behaviour'. I've used the following code:

```
    {ok, _, Reference} = compile:file("ssh", [binary]),

    false =
        lists:all(fun(_) ->
                     {ok, _, Testing} = compile:file("ssh", [binary]),
                     Testing =:= Reference
                  end,
                  lists:seq(1, 100)),
```

Meanwhile, it can be reproduced easily when using `erlc` from shell with the same files:

```
## the following files are from lib/ssh/src/
$ ls
ssh.erl  ssh.hrl  ssh_connect.hrl
$ for a in $(seq 1 100); do $ERL_TOP/bin/erlc ssh.erl; md5=$(md5sum ssh.beam | awk '{print $1}'); mv ssh.beam ssh.beam.${md5}; done
$ ls
ssh.beam.0c2b4dbfa37b12c2478782670083c40b  ssh.beam.6e58cd3ec2f0ba1de51959191be6a128  ssh.erl  ssh.hrl	ssh_connect.hrl
```

Any ideas, advices, etc.?

**Comment by garazdawi** [comment]
The behaviour relies on two different VMs being started when compiling so that the atom indexes end up being different for the metadata. You can use the `?CT_PEER` macro to start a new node in your testcase.

**Comment by delitrem** [comment]
> You can use the `?CT_PEER` macro to start a new node in your testcase.

Thank you very much for you hint! It helped.

I will test a little bit more and then commit a test case into my branch.

**Comment by delitrem** [comment]
Pushed changes to the `compile.erl` and test suite.

File: lib/compiler/test/compile_SUITE.erl [hunk 1]
```diff
@@ -1803,6 +1805,29 @@ deterministic_paths_1(DataDir, Name, Opts) ->
         file:set_cwd(Cwd)
     end.
 
+%% The test case uses ssh.erl from ssh application.
+deterministic_docs(Config) when is_list(Config) ->
+    DataDir = proplists:get_value(data_dir, Config),
+    Filepath = filename:join(DataDir, "ssh"),
+    false = deterministic_docs_1(Filepath, [binary], 100),
```
Comment (garazdawi) [inline lib/compiler/test/compile_SUITE.erl:1812]: Are 100 iterations needed to reliably trigger the fault? Or can it be reduced to limit runtime of the test?

Comment (delitrem) [reply lib/compiler/test/compile_SUITE.erl:1812]: > Are 100 iterations needed to reliably trigger the fault? Or can it be reduced to limit runtime of the test?

Sure, number of iterations can be reduced, but chances of the first case (which is without 'deterministic' option) failure will increase.

(after some testing)
With 10 iterations is fails sometimes.
It seems, optimal number of is 25, no failures at least on my local machine.

Comment (delitrem) [reply lib/compiler/test/compile_SUITE.erl]: > It seems, optimal number of is 25, no failures at least on my local machine.

Changed number of iterations to 25.

Comment (garazdawi) [reply lib/compiler/test/compile_SUITE.erl]: Great, thanks!

**Comment by delitrem** [comment]
Thank you very much for your assistance!

**Pull request merged**